circles-app/src/lib/components/markdown/ast.ts

```
@@ -0,0 +1,374 @@
+
+const allowedProtocols = new Set(['http:', 'https:', 'mailto:']);
+
+export function sanitizeUrl(raw: string): string | null {
+    const isString = typeof raw === 'string';
+    const input = isString ? raw : String(raw);
+    const trimmedOuter = input.trim();
+
+    const hasAngleBrackets = trimmedOuter.startsWith('<') && trimmedOuter.endsWith('>');
+    const candidate = hasAngleBrackets ? trimmedOuter.slice(1, -1).trim() : trimmedOuter;
+
+    const hasControlChars = /[\u0000-\u001F\u007F]/.test(candidate);
```

Biome (2.1.2)

[error] Unexpected control character in a regular expression.

Control characters are unusual and potentially incorrect inputs, so they are disallowed.

(lint/suspicious/noControlCharactersInRegex)